h/services/annotation_sync.py

```diff
@@ -53,11 +53,15 @@ def sync(self, limit):
                 # annotation is already present and up-to-date in Elasticsearch.
                 counter.annotation_synced(counter.Result.SYNCED_FORCED, job)
                 counter.job_completed(counter.Result.COMPLETED_FORCED, job)
-            elif not annotation_from_db:
+            elif (not annotation_from_db) and annotation_from_es:
                 # The annotation isn't in the DB or is marked as
-                # Annotation.deleted=True in the DB so we can't index it into
-                # Elasticsearch.
-                # (This method doesn't support deleting annotations from Elasticsearch yet.)
+                # Annotation.deleted=True in the DB but it still present in
+                # Elasticsearch. Delete the annotation from Elasticsearch.
+                counter.annotation_deleted(counter.Result.SYNCED_DELETED, job)
```

Comment:
I think you've spotted a pre-existing bug in the code here, although I believe the bug is inconsequential: the implementation of force mode kind of assumes that the annotation does exist in the DB, fails to account for `force=True` jobs whose annotation has been deleted from the DB.

## When `force=True` is used

For context, `force=True` is only used by the admin pages. Specifically, it's used when:

* The `/admin/search` page is used re-index all of a given user's annotations, if the admin checks the **Force** checkbox: https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/views/admin/search.py#L57
* The same admin page is used to re-index all of a _group_'s annotations, again if the admin checks the **Force** checkbox: https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/views/admin/search.py#L83
* When the `/admin/documents` page is used to move annotations from one URL to another. This one is a little less direct: the admin pages view [schedules](https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/views/admin/documents.py#L42) a Celery task which [calls](https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/tasks/url_migration.py#L8) `URLMigrationService.move_annotations_by_url()` which makes calls to `URLMigrationService.move_annotations()` which [calls](https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/services/url_migration.py#L80) `AnnotationWriteService.update_annotation()` with `update_timestamp=False`. When called with `update_timestamp=False` `AnnotationWriteService.update_annotation()` will [add jobs with `force=True`](https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/services/annotation_write.py#L144-L153). I believe this admin page is the one and only place where `update_timestamp=False` is used.
* When the `/admin/nipsa` page is used to NIPSA or un-NIPSA a user: <https://github.com/hypothesis/h/blob/45ce0b6248cf054f6d806bc0ee7f0eaebd54c4a4/h/services/nipsa.py#L77>

How to test `force=True` jobs for deleted annotations
-----------------------------------------------------

To test what happens when there's a `force=True` job whose annotation has already been deleted from the DB:

1. Login as `devdata_admin`
2. Go to http://localhost:5000/docs/help and make an annotation. This will add a `"sync_annotation"` job to the queue
3. Run h-periodic and wait for the sync_annotations task to clear the job from the queue
4. Stop h-periodic
5. Go to http://localhost:5000/admin/search and reindex all of `devdata_admin`'s annotations, checking the **Force** checkbox. This will add another `"sync_annotation"` job to the queue, this time with `force=True`
6. Delete the annotation. This will add _another_ `"sync_annotation"` job for the same annotation to the queue, this one with `force=False`
7. Start h-periodic again

What happens when there's a `force=True` job for a deleted annotation
---------------------------------------------------------------------

1. `AnnotationSyncService.sync()` will add the annotation to `counter.annotation_synced()` with `counter.Result.SYNCED_FORCED` as the metric and will add the job to `counter.job_completed` with `counter.Result.COMPLETED_FORCED` as the metric: https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/services/annotation_sync.py#L54-L55
2. Further down in `AnnotationSyncService.sync()` it'll pass the job to `JobQueueService.delete()`: https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/services/annotation_sync.py#L80
4. And it'll pass the annotation's ID to `BatchIndexer.index()`: https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/services/annotation_sync.py#L82-L83
5. `BatchIndexer.index()` [calls](https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/search/index.py#L45) a `_filtered_annotations()` helper method to fetch all the annotations  for the annotation IDs it was given. This [`_filtered_annotations()`](https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/search/index.py#L122-L130) function can only return annotations that still exist in the DB (obviously) so annotation IDs for annotations that've been purged from the DB are filtered out. The query also includes [a filter](https://github.com/hypothesis/h/blob/7554c4ba122b8ccde193b5ac7d8d8c396ae2b738/h/search/index.py#L135) that filters out annotations with `Annotation.deleted=True` (annotations that've been marked as deleted in the DB but not yet expunged by the hourly `purge_annotations()` task). So in either case deleted annotations are filtered out by `BatchIndexer`, it does not index them into Elasticsearch.

So the worst that's happening here is that `AnnotationSyncService.sync()` is reporting the `counter.Result.SYNCED_FORCED` when in fact, because the annotation was deleted, nothing was actually synced to Elasticsearch for this annotation.

Arguably `AnnotationSyncService.sync()` should be fixed here. At the very least it should not increment the `SYNCED_FORCED` metric when the annotation has been deleted from the DB. Perhaps it should even forceably re-delete the annotation from Elasticsearch? (And report a new `DELETED_FORCED` metric.)

`AnnotationSyncService`'s `DBHelper` and `ESHelper` contain optimisations to not fetch the annotations of `force=True` jobs from the DB or from Elasticsearch, since they aren't needed. If we wanted to make the change above I think we'd have to remove these (unimportant) optimisations because we'd need to know whether the annotations of `force=True` jobs have been deleted.

In practice, I'm gonna vote to ignore this and not making any changes here. I think the effects of this bug are minimal and fixing it now would be time-consuming and risky, and I don't have much time left to ship this PR.